### PR TITLE
add nm_join_bayes

### LINKED ISSUE #46: port run_sims helper from example project
[ This issue just a stub at this point. ]

The plan is now for something along the lines of the `run_sims` helper, which relies on mrgsolve and furrr, to live in bbr.bayes.

Implementing this should wait at least until PR #43 lands.

### LINKED ISSUE #50: `nm_join_bayes()`
* This could essentially be calling `bbr::nm_join()` on each chain sub-model, then row-binding them together, and summarize across chains.
  * you would end up with `input_rows` rows
  * be able to pass a `.point_fn = c("median", "mean")` to summarize by
* Put something in the docs that says "this is the fast way to do this, the better way is `@seealso run_sims()`"

### LINKED ISSUE #96: nm_join_bayes: is use of fixed table okay?
cc @seth127 @timwaterhouse @kylebaron

I'm wondering about the interaction of these two directions we decided to take with gh-75:

-   `submit_model` now tacks on its own table record for `run_sims` underneath so that we don't need to worry about doing extensive checking of a user-controlled table record.
    
    [9709c43](https://github.com/metrumresearchgroup/bbr.bayes/commit/9709c43f144295a6b8af10c07699b2f1e69d306d) (nmbayes submit\_model: write table for run\_sims, gh-75).

-   `run_sims` was renamed to `nm_join_bayes` and an `nm_join_bayes_quick` variant was added.
    
    [f2622db](https://github.com/metrumresearchgroup/bbr.bayes/commit/f2622db7567d55abd3cf4d82b1aeb82ca7119751) (nmbayes: rename run\_sims to nm\_join\_bayes, gh-75)
    
    [a100218](https://github.com/metrumresearchgroup/bbr.bayes/commit/a100218007240c25d8faa9fe9d6980f7c177d652) (nmbayes: export 'quick' wrapper around nm\_join\_bayes, gh-75).

`bbr::nm_join` supports specifying table file(s) via its `.files` argument. Given the new name, is it a problem that `nm_join_bayes` only reads in a behind-the-scenes table with a fixed set of columns?

### LINKED ISSUE #100: Problem with simulating in superset mode?
_This is a bit ill-defined right now, but hopefully the discussion will flesh it out..._

Discussion in issue #96 seems to be pointing towards supporting the `.superset` argument from `bbr::nm_join()` in `bbr.bayes::nm_bayes_join()` (possible resolution in [this comment](https://github.com/metrumresearchgroup/bbr.bayes/issues/96#issuecomment-1741271228)). Implementation-wise, this should be fine, but there was an outstanding question (originally from @kyleam ) about whether this could cause downstream problems with the `mrgsolve` model.

Essentially, the case in question would be:

* User calls `nm_join_bayes()` with `.superset = TRUE`
* This causes the original model data to come in, complete with any rows that had been ignored by NONMEM. Any tabled columns will be `NA`
* This gets passed to `mrgsolve` to simulate some new columns

The question is: will these rows that were ignored by NONMEM cause any problems for `mrgsolve`? Or even if they don't cause errors, could it return potentially misleading data, such that we should catch this and not let the user do it? Any thoughts from @kylebaron or anyone else are most appreciated.

### PR DESCRIPTION
**Pull request by kyleam**
When working on `run_sims` in gh-75, we decided to focus on the return value and rename the function to `nm_join_bayes`, treating it as the nmbayes analog of `nm_join`.  Going that direction led to gh-96, and consensus from discussion there was that 1) we should try to align with `nm_join` as closely as possible (e.g., support `.files`) and 2) ideally `nm_join` would be used underneath to avoid repeating some its more complicated logic.

This series implements that proposal and supersedes gh-75.  Most notable changes from gh-75:

 * Use `nm_join` underneath.

 * Don't write behind-the-scenes table when submitting model; instead use the `nm_join` default for the tables and allow callers to specify file via `.files`.

 * Drop argument for specifying custom data; instead add an argument that allows callers to specify a function that can be used to transform the data after the table-to-data join but before simulating with mrgsolve.

 * Support `.superset = TRUE`.

 * Drop `log_dv` argument; see discussion in gh-97.  (I'll follow up there.)

 * Flesh out tests.  At some point, @seth127 and I think it'd be nice to have the SME-inspected tests along the lines of what's discussed in gh-75, but we're not sure what that will look like, and it's likely to live outside this test suite, so we're not considering it a blocker for merging this PR.

Closes #46.
Closes #50.
Closes #96.
Closes #100.

Re: #97

cc: @kylebaron @timwaterhouse

---

- [x] Open PR `nm_join` "origin attribute" PR for bbr.  (This PR introduces handling for that feature but also has fallback code for older bbr versions.)
     update: https://github.com/metrumresearchgroup/bbr/pull/617

### CODE REVIEW

File: R/nm-join-bayes.R [hunk 1]
```diff
@@ -0,0 +1,583 @@
...
+#'   (e.g., for LOO calculations), not just the summary returned by this
+#'   function.
+#' @param ewres_npde Whether to replace EWRES and NPDE values obtained from
+#'   table with ones generated with \pkg{npde}.
+#' @param npde_decorr_method Pass this value to `decorr.method` of
+#'   [npde::autonpde()].
+#' @param presim_fn Before simulating, apply this function to the data frame
+#'   that results from joining the input data and table values. The main purpose
+#'   of this argument is to provide a way to do any column renames that are
+#'   required for the mrgsolve simulation (e.g., renaming a column to "TIME").
```
Comment (kyleam) [inline R/nm-join-bayes.R:84]: I don't love the name `presim_fn`.   If anyone has better ideas, let me know.

File: R/nm-join-bayes.R [hunk 2]
```diff
@@ -0,0 +1,583 @@
...
+  dfs[[1]] <- join(sdirs[1])
+  verbose_msg("\nRepeating join for other chains and combining results.")
+  withr::with_options(list(bbr.verbose = FALSE), {
+    for (i in 2:nchains) {
+      dfs[[i]] <- join(sdirs[i])
+    }
+  })
+
+  # nm_join_origin records the source of different columns.
+  origin <- attr(dfs[[1]], "nm_join_origin")
```
Comment (kyleam) [inline R/nm-join-bayes.R:338]: This is the bit that will require a PR on bbr's side.

File: R/nm-join-bayes.R [hunk 3]
```diff
@@ -0,0 +1,583 @@
...
+  }
+
+  return(list(data = data, tab = tab_sum))
+}
+
+call_presim <- function(fn, x, join_col) {
+  nrow_orig <- nrow(x)
+  x <- fn(x)
+
+  if (!inherits(x, "data.frame")) {
```
Comment (seth127) [inline R/nm-join-bayes.R:375]: These checks all look good to me. We're kind of getting into deep water here with letting the user pass in a function, but if the results passes these checks... I guess I think we're doing our due diligence.

**Review by seth127 — CHANGES_REQUESTED** [review]
This is all looking great to me. I left some comments, but they're all documentation things. 

The core code here has been reviewed by myself and various scientific SMEs several times now, so I think we're good to merge this now (pending docs edits). After that, we'll run it through the paces on some more realistic models and see if we come up with anything we want to change before the initial release.

Great work here.

File: R/nm-join-bayes.R [hunk 4]
```diff
@@ -0,0 +1,583 @@
...
+  checkmate::assert_string(dv_col)
+  checkmate::assert_string(y_col)
+  checkmate::assert_string(ipred_path, null.ok = TRUE)
+  checkmate::assert_numeric(probs, lower = 0, upper = 1, len = 2)
+  checkmate::assert_int(n_post)
+  npde_decorr_method <- match.arg(npde_decorr_method)
+  checkmate::assert_int(min_batch_size)
+
+  if (!is.null(.files) && any(fs::is_absolute_path(.files))) {
+    stop("`.files` must be relative paths.")
```
[the remark this answers is not in the document]
Comment (kyleam) [reply R/nm-join-bayes.R:143]: Good catch.  Added in 1778067.

File: R/nm-join-bayes.R [hunk 5]
```diff
@@ -0,0 +1,583 @@
...
+    }
+  }
+
+  if (isTRUE(ewres_npde)) {
+    if (!isTRUE(epred)) {
+      stop("`ewres_npde = TRUE` depends on `epred = TRUE`.")
+    }
+
+    if (!requireNamespace("npde", quietly = TRUE)) {
+      stop("`npde = TRUE` requires npde package.")
```
[the remark this answers is not in the document]
Comment (kyleam) [reply R/nm-join-bayes.R]: Doh, fixed in a36551d.

File: R/nm-join-bayes.R [hunk 6]
```diff
@@ -0,0 +1,583 @@
...
+#' the `nm_join_bayes()` call inside a [progressr::with_progress()] call.
+#'
+#' @inheritParams bbr::nm_join
+#'
+#' @param .mod A `bbi_nmbayes_model` object.
+#' @param mod_mrgsolve An mrgsolve model object, potentially updated to be
+#'   optimized for simulation from the data set and model (e.g., ODE solver
+#'   tolerance). This must capture `y_col`.
+#' @param .join_col Use this column to join tables files to the input data.
+#' @param .files Paths to table files to pass to [bbr::nm_join()] calls (on per
```
[the remark this answers is not in the document]
Comment (kyleam) [reply R/nm-join-bayes.R]: fixed in 3cbf3c5

**Comment by kyleam** [comment]
@seth127 Thanks for the review.   Pushed updates.   Back to you.

**Review by seth127 — APPROVED** [review]
Looking great. Thanks again for all the work on this.

**Pull request merged**